### PR TITLE
connect: Add AWS PCA provider

### PR DESCRIPTION
**Pull request by banks**
**This is built on top of #6775**

This adds support for AWS ACM Private CA as a Connect CA Provider.

There are a few limitations of PCA that will need to be documented 
 - it can't blindly cross-sign other root certs which our CA interface relies on for rotations. 
 - it only supports a few key types and key lengths
 - it can't sign certs that are valid for less than a day
 - there are strict rate limits and limits on the number of CAs accounts can create.

The tests added require an ENV var setup and we need to build a suitable CI runner that doesn't run all the time since PCA is relatively expensive and slow.

But here is it running locally against AWS PCA.

```bash
$envchain csl go test -v github.com/hashicorp/consul/agent/connect/ca -run '^TestAWS'
=== RUN   TestAWSBootstrapAndSignPrimary
=== RUN   TestAWSBootstrapAndSignPrimary/Default_Key_Type_(EC_256)
=== RUN   TestAWSBootstrapAndSignPrimary/RSA_2048
--- PASS: TestAWSBootstrapAndSignPrimary (34.50s)
    --- PASS: TestAWSBootstrapAndSignPrimary/Default_Key_Type_(EC_256) (17.45s)
    --- PASS: TestAWSBootstrapAndSignPrimary/RSA_2048 (17.05s)
=== RUN   TestAWSBootstrapAndSignSecondary
--- PASS: TestAWSBootstrapAndSignSecondary (51.19s)
=== RUN   TestAWSBootstrapAndSignSecondaryConsul
=== RUN   TestAWSBootstrapAndSignSecondaryConsul/pri=consul,sec=aws
=== RUN   TestAWSBootstrapAndSignSecondaryConsul/pri=aws,sec=consul
--- PASS: TestAWSBootstrapAndSignSecondaryConsul (28.75s)
    --- PASS: TestAWSBootstrapAndSignSecondaryConsul/pri=consul,sec=aws (11.62s)
    --- PASS: TestAWSBootstrapAndSignSecondaryConsul/pri=aws,sec=consul (17.13s)
PASS
ok  	github.com/hashicorp/consul/agent/connect/ca	116.222s
```

# TODO

 - [x] Handle rate limiting on Sign nicely - plumb it through to the CA endpoint so clients will treat it the same way as built-in consul rate limiting and retry with a backoff
 - [x] Write Docs!
 - [x] With @alvin-huang figure out how to put this in CI (possibly a separate PR).

# Future Work

 - **Support the Primary CA being a subordinate CA of some other root** -  we should support this but there are a few places in the code so far I've left a TODO here as it needs a bit more sophistication.
 - **Refactor CA Interface** - It occured to me while deep in the weeds here that the current interface is both more complex than it needs to be and assumes a lot about CA capabilities - i.e. that they support blind cross-signing a bare certificate from another Root without a valid CSR from that root. I think before we work on further CA integrations we should simplify this:
    - Have the Primary DC always use a separate Intermediate even when it generated the root
    - Have the standard workflow essentially be the same as the SecondaryDC workflow now where each DC generates an intermediate to sign with and then get's it's CSR signed _twice_ once by it's own root and once by the previous root (this would be possible with AWS PCA and most other CAs and would result in a cross-signed intermediate that links back to either root)

### CODE REVIEW

File: agent/connect/ca/provider_aws.go [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+		select {
+		case <-a.stopCh:
+			// Provider discarded
+			a.logger.Print("[WARN] connect.ca.aws: provider instance terminated"+
+				" while waiting for a new certificate %s to be ready.",
+				issueOutput.CertificateArn)
+			return "", fmt.Errorf("provider terminated")
+		case <-time.After(a.config.PollInterval):
+			// Continue looping...
+		}
```
Comment (kyhavlov) [inline agent/connect/ca/provider_aws.go]: I assume this request doesn't time out to AWS in practice but this could hang indefinitely if the request status stalled on their end for some reason, right? Maybe we should add a hard timeout here to fail the request after enough waiting. Ditto for the above CA request loop.

File: agent/connect/ca/provider_aws.go [hunk 2]
```diff
@@ -0,0 +1,656 @@
...
+	}
+
+	if a.config.DeleteOnExit {
+		if err := a.disablePCA(); err != nil {
+			// Log the error but continue trying to delete as some errors may still
+			// allow that and this is best-effort delete anyway.
+			a.logger.Printf("[ERR] connect.ca.aws: failed to disable PCA %s: %s",
+				a.arn, err)
+		}
+		if err := a.deletePCA(); err != nil {
```
Comment (kyhavlov) [inline agent/connect/ca/provider_aws.go]: This couldn't be called on the new active PCA after a root rotation, right? I *think* the new root being different and triggering a rotation should mean it can't be the same PCA, but just wanted to ask.

File: agent/connect/ca/provider_aws.go (hunk 1 above)
Comment (kyhavlov) [reply agent/connect/ca/provider_aws.go]: I couldnt find anything in the AWS docs for these endpoints about having a hard timeout, and even if they do we might still want a cutoff on our end in case the status of the request stalls for some reason.

Comment (banks) [reply agent/connect/ca/provider_aws.go]: Yeah this is a great call. Our own clients probably don't expect to wait that long although net/rpc doesn't actually have timeouts right now.

I think 30 seconds to sign a cert is probably suficient. I'm also thinking of getting rid of the fixed poll interval and instead do something like:
 - wait 20ms before first attempt since I've never seen an immediate one work first time
 - wait 500ms before second
 - wait 1 second before third, (hopefully majority complete this fast - i've always seen it work after first 5 second wait so far)
 - wait 3 seconds before fourth
 - wait 5 seconds for the next 5 tries, total time after that is ~30 seconds so just give up and return a timeout error. We can tweak these later, but having them user-configurable doesn't make much sense to me for now. Maybe the overall timeout might be configurable later.

File: agent/connect/ca/provider_aws.go (hunk 2 above)
Comment (banks) [reply agent/connect/ca/provider_aws.go]: It should never be called outside of tests - we won't document the `DeleteOnExit` config as that is only for tests really. In prod you'd never want CA resources deleted like this right after rotation like you said.

**Comment by banks** [comment]
@kyhavlov thanks! I've addressed the things mentioned and the rate-limit TODO.

Still battling with testing - I've hit the limit on how many Deleted CAs you can have which is not what is documented so we need to figure that out too before we can enable in CI.

I still need to do docs but I think this is good to go from Code perspective!

File: agent/connect/ca/provider_aws.go [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+}
+
+// Configure implements Provider
+func (a *AWSProvider) Configure(cfg ProviderConfig) error {
+	config, err := ParseAWSCAConfig(cfg.RawConfig)
+	if err != nil {
+		return err
+	}
+
+	// We only support setting IAM credentials through the normal methods ENV,
```
Comment (rboyer) [inline agent/connect/ca/provider_aws.go]: Would it be possible to have the state store diverge if all of the servers had different IAM credentials that implied that they were manipulating CAs in wildly different regions from each other?

Comment (banks) [reply agent/connect/ca/provider_aws.go]: No it would just break - the actual state stored is the ARN of a specific CA which only exists in one region so if servers have different IAM configs then some just would fail to load it.

File: agent/connect/ca/provider_aws.go [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+	}
+
+	// If we are in a secondary DC this is all we can do for now - the rest is
+	// handled by the Initialization routine of calling GenerateIntermediateCSR
+	// and then SetIntermediate.
+	if !a.isPrimary {
+		return nil
+	}
+
+	// CA is created and in PENDING_CERTIFCATE state, generate a self-signed cert
```
Comment (rboyer) [inline agent/connect/ca/provider_aws.go]:
```suggestion
	// CA is created and in PENDING_CERTIFICATE state, generate a self-signed cert
```

**Comment by banks** [comment]
Rebased on master now parent PR is merged. Seems clean will watch CI...

File: agent/structs/connect_ca.go [hunk 5]
```diff
@@ -460,6 +461,13 @@ type VaultCAProviderConfig struct {
 	TLSSkipVerify bool
 }
 
+type AWSCAProviderConfig struct {
```
Comment (rboyer) [inline agent/structs/connect_ca.go]: I think you are missing the mirror change in `api`.

File: agent/connect/ca/provider_aws.go [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+			SigningAlgorithm: aws.String(signAlg),
+		},
+		RevocationConfiguration: &acmpca.RevocationConfiguration{
+			// TODO support CRL in future when we manage revocation in Connect more
+			// generally.
+			CrlConfiguration: &acmpca.CrlConfiguration{
+				Enabled: aws.Bool(false),
+			},
+		},
+		// uid is unique each PCA we create so use it as an idempotency string. We
```
Comment (rboyer) [inline agent/connect/ca/provider_aws.go]:
```suggestion
		// uid is unique to each PCA we create so use it as an idempotency string. We
```

File: agent/connect/ca/provider_aws.go [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+	if err != nil {
+		return err
+	}
+
+	// wait for PCA to be created
+	newARN := *createOutput.CertificateAuthorityArn
+	describeInput := acmpca.DescribeCertificateAuthorityInput{
+		CertificateAuthorityArn: aws.String(newARN),
+	}
+	_, err = a.pollLoop("Private CA", 2*time.Minute, func() (bool, string, error) {
```
Comment (rboyer) [inline agent/connect/ca/provider_aws.go]: Maybe extract this to a more visible constant?

File: agent/connect/ca/provider_aws.go [hunk 8]
```diff
@@ -0,0 +1,700 @@
...
+		describeOutput, err := a.client.DescribeCertificateAuthority(&describeInput)
+		if err != nil {
+			if err.(awserr.Error).Code() != acmpca.ErrCodeRequestInProgressException {
+				a.logger.Printf("[ERR] connect.ca.aws: error describing PCA: %s", err.Error())
```
Comment (rboyer) [inline agent/connect/ca/provider_aws.go]: I thought it was a small antipattern to both log an error and also return the same error, due to the returned error almost definitely getting logged again one level up.

File: agent/connect/ca/provider_aws.go [hunk 9]
```diff
@@ -0,0 +1,700 @@
...
+}
+
+func (a *AWSProvider) getCACSR() (string, error) {
+	input := &acmpca.GetCertificateAuthorityCsrInput{
+		CertificateAuthorityArn: aws.String(a.arn),
+	}
+	a.logger.Printf("[DEBUG] connect.ca.aws: retrieving CSR for %s", a.arn)
+	output, err := a.client.GetCertificateAuthorityCsr(input)
+	if err != nil {
+		a.logger.Printf("[ERR] connect.ca.aws: error retrieving CSR: %s", err)
```
Comment (rboyer) [inline agent/connect/ca/provider_aws.go]: This is another case of logging and returning the same error which may cause duplication.

File: agent/connect/ca/provider_aws.go [hunk 10]
```diff
@@ -0,0 +1,700 @@
...
+		return "", err
+	}
+
+	issueInput := acmpca.IssueCertificateInput{
+		CertificateAuthorityArn: aws.String(a.arn),
+		Csr:                     []byte(csrPEM),
+		SigningAlgorithm:        aws.String(signAlg),
+		TemplateArn:             aws.String(templateARN),
+		Validity: &acmpca.Validity{
+			Value: aws.Int64(int64(ttl.Seconds() / 86400.0)),
```
Comment (rboyer) [inline agent/connect/ca/provider_aws.go]: nit: I know the stdlib explicitly does not provide a `time.Day` due to DST issues, but seeing "24 hours" in code may be somewhat more obvious than "86400 seconds" to folks who don't play with clocks all the time.
```
const roughDay = 24 * time.Hour
aws.Int64(int64(ttl / roughDay))
```

File: agent/connect/ca/provider_aws.go [hunk 11]
```diff
@@ -0,0 +1,700 @@
...
+		return "", fmt.Errorf("error issuing certificate from PCA: %s", err)
+	}
+
+	// wait for certificate to be created
+	certInput := acmpca.GetCertificateInput{
+		CertificateAuthorityArn: aws.String(a.arn),
+		CertificateArn:          issueOutput.CertificateArn,
+	}
+	return a.pollLoop(fmt.Sprintf("certificate %s", *issueOutput.CertificateArn),
+		45*time.Second,
```
Comment (rboyer) [inline agent/connect/ca/provider_aws.go]: Possibly consider uplifting this to a more visible package constant.

File: website/source/docs/connect/ca/aws.html.md [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+-> This page documents the specifics of the AWS ACM Private CA provider.
+Please read the [certificate management overview](/docs/connect/ca.html)
+page first to understand how Consul manages certificates with configurable
+CA providers.
+
+## Requirements
+
+The ACM Private CA Provider needs to be authorized via IAM credentials to
+perform operations. Every Consul Server needs to be running in an environment
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
perform operations. Every Consul server needs to be running in an environment
```

File: website/source/docs/connect/ca/aws.html.md [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+where a suitable IAM configuration is present.
+
+The [standard AWS SDK credential
+locations](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials)
+are used which means that suitable credentials and region configuration need to be present in one of:
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
are used, which means that suitable credentials and region configuration need to be present in one of the following:
```

File: website/source/docs/connect/ca/aws.html.md [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+ 1. Environment variables
+ 2. Shared credentials file
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
 1. Shared credentials file
```
Update based on eng style guide

File: website/source/docs/connect/ca/aws.html.md [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+ 3. Via an EC2 instance role
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
 1. Via an EC2 instance role
```

File: website/source/docs/connect/ca/aws.html.md [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+
+The ACM Private CA provider is enabled by setting the `ca_provider` to
+`"aws-pca"`. At this time there is only one, optional configuration value.
+
+```hcl
+connect {
+    enabled = true
+    ca_provider = "aws-pca"
+    ca_config {
+      existing_arn = "..."
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]: Can this be a real example or is it too complicated?

File: website/source/docs/connect/ca/aws.html.md [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+    }
+}
+```
+
+ ~> Note that suitable AWS IAM credentials are necessary for the provider to
+ work, however these are not configured in the Consul config which is typically
+ on disk and rely on the [standard AWS SDK configuration
+ locations](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials).
+
+The set of configuration options is listed below. The
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
The configuration options are listed below. Note, the
```

File: website/source/docs/connect/ca/aws.html.md [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+first key is the value used in API calls while the second key (after the `/`)
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
first key is the value used in API calls and the second key (after the `/`)
```

File: website/source/docs/connect/ca/aws.html.md [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+is used if configuring in an agent configuration file.
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
is used if you're adding configuring to the agent's configuration file.
```

File: website/source/docs/connect/ca/aws.html.md [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+
+  * `ExistingARN` / `existing_arn` (`string: <optional>`) - The Amazon Resource
+    Name (ARN) of an existing private CA in your ACM account. If specified,
+    Consul will attempt to use the existing CA to issue certificates. Currently
+    this **must be a root CA** in the primary datacenter although in the future
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
    this **must be a root CA** in the primary datacenter.

File: website/source/docs/connect/ca/aws.html.md [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+    we intend to support subordinate CAs. In a secondary datacenter, it must be
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
   In a secondary datacenter, it must be
```

File: website/source/docs/connect/ca/aws.html.md [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+    a subordinate CA who's root is the same as the root used in the primary
+    datacenter - if it's not Consul will create a new one from the Primary
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]: The if/thens are a little hard to follow. Maybe break them into bullet points?

File: website/source/docs/connect/ca/aws.html.md [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+      secondary DC that it will automatically get signed by the primary's root.
+
+## Limitations
+
+ACM Private CA has several
+[limits](https://docs.aws.amazon.com/acm-pca/latest/userguide/PcaLimits.html)
+that restrict how fast certificates can be issued. This may impact how quickly
+large clusters can rotate all issued certificates.
+
+At the current time the ACM Private CA provider for Connect has some additional
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
Currently, the ACM Private CA provider for Connect has some additional
```

File: website/source/docs/connect/ca/aws.html.md [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+limitations described below.
+
+### Unable to Cross-sign Other CAs
+
+It's currently not possible to cross-sign other CA provider's root certificates
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
It's not possible to cross-sign other CA provider's root certificates
```

File: website/source/docs/connect/ca/aws.html.md [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+during a migration. This is due to a ACM Private CA not having a mechanism to
+blindly cross-sign another root certificate without a CSR being generated. Both
+Consul's built in CA and Vault can do this and the current workflow for managing
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
Consul's built-in CA and Vault can do this and the current workflow for managing
```

File: website/source/docs/connect/ca/aws.html.md [hunk 26]
```diff
@@ -0,0 +1,146 @@
...
+CAs relies on it.
+
+In the future it should be possible to remove this limitation by changing the
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]: I'm alway hesitant to talk about future functionality, unless it's not in-progress work, since the future could be 2 months or 2 years.

File: agent/connect/ca/provider_aws.go (hunk 8 above)
Comment (banks) [reply agent/connect/ca/provider_aws.go]: It is, for some reason I thought this case warranted it because the top level one might not surface but now I look at it I'm not sure why that would be the case.

File: agent/structs/connect_ca.go (hunk 5 above)
Comment (banks) [reply agent/structs/connect_ca.go]: Right now we don't have structs in `api` for provider-specific configs since they are just a `map[string]interface{}` There is one for the built in provider but not sure why honestly.

The one thing I did miss there though in my last PR was adding the State field which is now in the response although can't be set in a request. I should also validate that the user isn't writing it directly in Config PUT too I suppose.

File: website/source/docs/connect/ca/aws.html.md [hunk 27]
```diff
@@ -0,0 +1,146 @@
+---
+layout: "docs"
+page_title: "Connect - Certificate Management"
```
[the remark this answers is not in the document]
Comment (banks) [reply website/source/docs/connect/ca/aws.html.md]: I guess so - I copied from Vault which sticks with this title. @kaitlincarter-hc do you think the title should change on these pages? That seems best to me but maybe there was a good reason to keep it the same on all of the provider pages?

File: website/source/docs/connect/ca/aws.html.md (hunk 19 above)
Comment (banks) [reply website/source/docs/connect/ca/aws.html.md]: I copied this from the Vault page so we should maybe update that to match too?

File: website/source/layouts/docs.erb [hunk 28]
```diff
@@ -546,6 +546,9 @@
               <li<%= sidebar_current("docs-connect-ca-vault") %>>
                 <a href="/docs/connect/ca/vault.html">Vault</a>
               </li>
+              <li<%= sidebar_current("docs-connect-ca-aws") %>>
+                <a href="/docs/connect/ca/aws.html">ACM Private CA</a>
```
[the remark this answers is not in the document]
Comment (banks) [reply website/source/layouts/docs.erb]: Me too, but I trying to respect the official naming for the product which is either "ACM Private CA" or "AWS Certificate Manger Private Certificate Authority" which is way too long for the side bar.

We could do something like `(AWS) ACM Private CA` or do you think `AWS ACM Private CA` is reasonable and we'll see if AWS have any concerns with that?

@kaitlincarter-hc any opinion?

File: website/source/docs/connect/ca/aws.html.md [hunk 29]
```diff
@@ -0,0 +1,146 @@
...
+Consul's built in CA and Vault can do this and the current workflow for managing
+CAs relies on it.
+
+In the future it should be possible to remove this limitation by changing the
+way Consul manages CA certificates significantly.
+
+For now, the limitation means that once ACM Private CA is configured as the CA
+provider, it is not possible to reconfigure a different CA provider, or rotate
+the root CA key without potentially observing some transient connection
+failures. See the section on [Forced rotation without
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
failures. See the section on [forced rotation without
```

File: website/source/docs/connect/ca/aws.html.md [hunk 30]
```diff
@@ -0,0 +1,146 @@
...
+cross-signing](/docs/connect/ca.html#forced-rotation-without-cross-signing) for
+more details.
+
+### Primary DC Must be a Root CA
+
+Currently, if an existing ACM Private CA is used, the primary DC must use a Root
+CA directly to issue certificates. We expect to remove this limitation in the
+future.
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]: see note above about "future"

File: website/source/docs/connect/ca/aws.html.md [hunk 31]
```diff
@@ -0,0 +1,146 @@
...
+
+To help estimate costs, an example is provided below of the resources that would
+be used.
+
+~> This is intended to illustrate the behavior of the CA for cost planning
+purposes. Please refer to the [pricing for ACM Private
+CA](https://aws.amazon.com/certificate-manager/pricing/) for actual cost
+information.
+
+Assume the following Consul datacenters exist and all will use ACM Private CA as
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
Assume the following Consul datacenters exist and all agents will use ACM Private CA as
```

File: website/source/docs/connect/ca/aws.html.md [hunk 32]
```diff
@@ -0,0 +1,146 @@
...
+cross-signing](/docs/connect/ca.html#forced-rotation-without-cross-signing) for
+more details.
+
+### Primary DC Must be a Root CA
+
+Currently, if an existing ACM Private CA is used, the primary DC must use a Root
+CA directly to issue certificates. We expect to remove this limitation in the
+future.
+
+## Cost Example
```
Comment (kaitlincarter-hc) [inline website/source/docs/connect/ca/aws.html.md]:
```suggestion
## Cost Planning 
```

File: go.mod [hunk 33]
```diff
@@ -14,6 +14,7 @@ require (
 	github.com/armon/circbuf v0.0.0-20150827004946-bbbad097214e
 	github.com/armon/go-metrics v0.0.0-20190430140413-ec5e00d3c878
 	github.com/armon/go-radix v1.0.0
+	github.com/aws/aws-sdk-go v1.25.32
```
[the remark this answers is not in the document]
Comment (banks) [reply go.mod]: Done. It works fine against our Consul cluster in Dev AWS account

File: website/source/docs/connect/ca/aws.html.md (hunk 31 above)
Comment (banks) [reply website/source/docs/connect/ca/aws.html.md]: It's really the datacenter rather than the agents that use ACM since CA provider is configured per DC. Does it read better as:

> Assume the following Consul datacenters exist and are configured to use ACM Private CA...

File: agent/structs/connect_ca.go (hunk 5 above)
Comment (banks) [reply agent/structs/connect_ca.go]: ^ PUT validation for no state modification done locally.

File: website/source/docs/connect/ca/aws.html.md (hunk 16 above)
Comment (banks) [reply website/source/docs/connect/ca/aws.html.md]: Didn't want to use a real one because it's long and contains potentially sensitive info like our AWS account ID. But I'll swap it for an example one from AWS docs and see how that looks:
`arn:aws:acm-pca:region:account:certificate-authority/12345678-1234-1234-123456789012`

File: website/source/docs/connect/ca/aws.html.md (hunk 22 above)
Comment (banks) [reply website/source/docs/connect/ca/aws.html.md]: Attempted to clean it up a bit. See what you think.

Comment (banks) [reply website/source/docs/connect/ca/aws.html.md]: image

File: website/source/docs/connect/ca/aws.html.md (hunk 26 above)
Comment (banks) [reply website/source/docs/connect/ca/aws.html.md]: That's fair. I'll remove those comments.

**Comment by banks** [comment]
@kaitlincarter-hc @rboyer thanks so much for the reviews! I think I've addressed everything other than those two open naming/titling questions. Let me know what you think!

File: agent/connect_ca_endpoint.go [hunk 34]
```diff
@@ -69,5 +70,11 @@ func (s *HTTPServer) ConnectCAConfigurationSet(resp http.ResponseWriter, req *ht
 
 	var reply interface{}
 	err := s.agent.RPC("ConnectCA.ConfigurationSet", &args, &reply)
+	if err != nil && err.Error() == consul.ErrStateReadOnly.Error() {
+		return nil, BadRequestError{
+			Reason: "Provider State is read-only it must be omitted" +
```
Comment (rboyer) [inline agent/connect_ca_endpoint.go]:
```suggestion
			Reason: "Provider State is read-only. It must be omitted" +
```

File: agent/connect/ca/provider_aws.go [hunk 35]
```diff
@@ -0,0 +1,711 @@
...
+	a.isPrimary = cfg.IsPrimary
+	a.clusterID = cfg.ClusterID
+	a.datacenter = cfg.Datacenter
+	a.client = acmpca.New(awsSession)
+	a.stopCh = make(chan struct{})
+
+	// Load the ARN from config or previous state.
+	if config.ExistingARN != "" {
+		a.arn = config.ExistingARN
+	} else if arn := cfg.State["CA_ARN"]; arn != "" {
```
Comment (rboyer) [inline agent/connect/ca/provider_aws.go]: trivial: maybe consider making this key a constant
